File: src/schublib.cpp
/*  Littlewood-Richardson Calculator
 *  Copyright (C) 1999- Anders S. Buch (asbuch at math rutgers edu)
 *  See the file LICENSE for license information.
 */

#include "lrcalc/schublib.hpp"

#include <assert.h>
#include <stdint.h>

#include <limits>
#include <new>
#include <utility>

#include "lrcalc/cpp_lib.hpp"
#include "lrcalc/ivector.hpp"
#include "lrcalc/ivlincomb.hpp"
#include "lrcalc/perm.hpp"

static void _trans(ivector* w, int vars, ivlincomb* res);

ivlincomb* trans(ivector* w, int vars)
{
	ivlc_ptr res = ivlc_create();
	_trans(w, vars, res.get());
	return res.release();
}

static void _trans(ivector* w, int vars, ivlincomb* res)
{
	ivlc_reset(res);

	uint32_t nw = iv_length(w);
	int n = perm_group(w);
	w->length = uint32_t(n);

	int r = n - 1;
	while (r > 0 && iv_elem(w, r - 1) < iv_elem(w, r)) r--;
	if (r <= 0)
	{
		ivector* xx = iv_new_zero(vars ? uint32_t(vars) : 1);
		w->length = nw;
		ivlc_insert(res, xx, iv_hash(xx), 1);
		return;
	}
	if (vars < r) vars = r;

	int s = r + 1;
	while (s < n && iv_elem(w, r - 1) > iv_elem(w, s)) s++;

	int wr = iv_elem(w, r - 1);
	int ws = iv_elem(w, s - 1);

	ivector* v = w;
	iv_elem(v, s - 1) = wr;
	iv_elem(v, r - 1) = ws;

	ivlincomb* tmp = trans(v, vars);
	for (auto& kv : ivlc_iterator(tmp))
	{
		ivector* xx = kv.key;
		iv_elem(xx, r - 1)++;
		uint32_t hash = iv_hash(xx);
		ivlc_insert(res, xx, hash, kv.value);
	}

	int last = 0;
	int vr = iv_elem(v, r - 1);
	for (int i = r - 1; i >= 1; i--)
	{
		int vi = iv_elem(v, i - 1);
		if (last < vi && vi < vr)
		{
			last = vi;
			iv_elem(v, i - 1) = vr;
			iv_elem(v, r - 1) = vi;
			_trans(v, vars, tmp);
			ivlc_add_multiple(res, 1, tmp, LC_FREE_ZERO);
			iv_elem(v, i - 1) = vi;
		}
	}

	w->length = nw;
	iv_elem(w, s - 1) = ws;
	iv_elem(w, r - 1) = wr;
	ivlc_free(tmp);
}

static void _monk_add(uint32_t i, const ivlc_ptr& slc, int rank, ivlc_ptr& res)
{
	for (const auto& kv : ivlc_iterator(slc))
	{
		const ivector* w = kv.key;
		int c = kv.value;
		uint32_t n = iv_length(w);
		int wi = (i <= n) ? iv_elem(w, i - 1) : int(i);

		if (i <= n + 1)
		{
			int last = 0;
			uint32_t ulen = (i > n) ? i : n;
			for (uint32_t j = i - 1; j >= 1; j--)
				if (last < iv_elem(w, j - 1) && iv_elem(w, j - 1) < wi)
				{
					last = iv_elem(w, j - 1);
					ivector* u = iv_new(ulen);
					for (uint32_t t = 0; t < n; t++) iv_elem(u, t) = iv_elem(w, t);
					for (uint32_t t = n; t < ulen; t++) iv_elem(u, t) = int(t + 1);
					iv_elem(u, j - 1) = wi;
					iv_elem(u, i - 1) = last;
					ivlc_add_element(res.get(), -c, u, iv_hash(u), LC_FREE_ZERO);
				}
		}
		else
		{
			ivector* u = iv_new(i);
			for (uint32_t t = 0; t < n; t++) iv_elem(u, t) = iv_elem(w, t);
			for (uint32_t t = n; t < i - 2; t++) iv_elem(u, t) = int(t + 1);
			iv_elem(u, i - 2) = int(i);
			iv_elem(u, i - 1) = int(i) - 1;
			ivlc_add_element(res.get(), -c, u, iv_hash(u), LC_FREE_ZERO);
		}

		if (i >= n + 1)
		{
			ivector* u = iv_new(i + 1);
			for (uint32_t t = 0; t < n; t++) iv_elem(u, t) = iv_elem(w, t);
			for (uint32_t t = n; t < i; t++) iv_elem(u, t) = int(t + 1);
			iv_elem(u, i - 1) = int(i) + 1;
			iv_elem(u, i) = int(i);
			ivlc_add_element(res.get(), c, u, iv_hash(u), LC_FREE_ZERO);
		}
		else
		{
			int last = std::numeric_limits<int>::max();
			for (uint32_t j = i + 1; j <= n; j++)
				if (wi < iv_elem(w, j - 1) && iv_elem(w, j - 1) < last)
				{
					last = iv_elem(w, j - 1);
					ivector* u = iv_new(n);
					for (uint32_t t = 0; t < n; t++) iv_elem(u, t) = iv_elem(w, t);
					iv_elem(u, i - 1) = last;
					iv_elem(u, j - 1) = wi;
					ivlc_add_element(res.get(), c, u, iv_hash(u), LC_FREE_ZERO);
				}
			if (last > int(n) && int(n) < rank)
			{
				ivector* u = iv_new(n + 1);
				for (uint32_t t = 0; t < n; t++) iv_elem(u, t) = iv_elem(w, t);
				iv_elem(u, i - 1) = int(n) + 1;
				iv_elem(u, n) = wi;
				ivlc_add_element(res.get(), c, u, iv_hash(u), LC_FREE_ZERO);
			}
		}
	}
}

static void _mult_ps(void** poly, uint32_t n, uint32_t maxvar, const ivector* perm, int rank, ivlc_ptr& res);

ivlincomb* mult_poly_schubert(ivlincomb* poly, ivector* perm, int rank)
{
	ivlc_ptr poly_ptr{poly};
	uint32_t n = ivlc_card(poly_ptr.get());
	if (n == 0) return poly;

	if (rank == 0) rank = std::numeric_limits<int>::max();

	auto p = new (std::nothrow) void*[2 * n];
	if (p == nullptr) return nullptr;
	uint32_t i = 0;
	uint32_t maxvar = 0;
	for (auto& kv : ivlc_iterator(poly_ptr))
	{
		ivector* xx = kv.key;
		uint32_t j = iv_length(xx);
		while (j > 0 && iv_elem(xx, j - 1) == 0) j--;
		xx->length = j;
		if (maxvar < j) maxvar = j;
		p[i++] = kv.key;
		p[i++] = reinterpret_cast<void*>(long(kv.value));
	}
	assert(i == 2 * n);
	ivlc_reset(poly_ptr.get());

	uint32_t svlen = iv_length(perm);
	perm->length = uint32_t(perm_group(perm));
	_mult_ps(p, n, maxvar, perm, rank, poly_ptr);
	perm->length = svlen;

	for (i = 0; i < n; i++) iv_free(static_cast<ivector*>(p[2 * i]));
	delete[] p;

	return poly_ptr.release();
}

static void _mult_ps(void** poly, uint32_t n, uint32_t maxvar, const ivector* perm, int rank, ivlc_ptr& res)
{
	if (maxvar == 0)
	{
		ivector* w = iv_new_copy(perm); /* FIXME: OPTIMIZE! */
		int c = int(reinterpret_cast<long>(poly[1]));
		ivlc_add_element(res.get(), c, w, iv_hash(w), LC_FREE_ZERO);
		return;
	}

	uint32_t mv0 = 0;
	uint32_t mv1 = 0;
	uint32_t j = 0;
	for (uint32_t i = 0; i < n; i++)
	{
		ivector* xx = static_cast<ivector*>(poly[2 * i]);
		uint32_t lnxx = iv_length(xx);
		if (lnxx < maxvar)
		{
			if (mv0 < lnxx) mv0 = lnxx;
		}
		else
		{
			iv_elem(xx, maxvar - 1)--;
			while (lnxx > 0 && iv_elem(xx, lnxx - 1) == 0) lnxx--;
			xx->length = lnxx;
			if (mv1 < lnxx) mv1 = lnxx;
			poly[2 * i] = poly[2 * j];
			poly[2 * j] = xx;
			std::swap(poly[2 * i + 1], poly[2 * j + 1]);
			j++;
		}
	}

	ivlc_ptr res1 = ivlc_create();
	_mult_ps(poly, j, mv1, perm, rank, res1);
	_monk_add(maxvar, res1, rank, res);

	if (j < n) _mult_ps(poly + 2 * j, n - j, mv0, perm, rank, res);
}

ivlincomb* mult_schubert(ivector* w1, ivector* w2, int rank)
{
	int w1len = perm_length(w1);
	int w2len = perm_length(w2);
	if (w1len > w2len)
	{
		std::swap(w1, w2);
		std::swap(w1len, w2len);
	}

	uint32_t svlen1 = iv_length(w1);
	uint32_t svlen2 = iv_length(w2);
	w1->length = uint32_t(perm_group(w1));
	w2->length = uint32_t(perm_group(w2));
	ivlincomb* lc = nullptr;

	if (rank == 0)
		rank = std::numeric_limits<int>::max();
	else if (2 * (w1len + w2len) > rank * (rank - 1) || bruhat_zero(w1, w2, rank))
	{
		lc = ivlc_new(IVLC_HASHTABLE_SZ, IVLC_ARRAY_SZ);
		goto free_return;
	}

	{
		ivlincomb* poly = trans(w1, 0);
		lc = mult_poly_schubert(poly, w2, rank);
	}

free_return:
	w1->length = svlen1;
	w2->length = svlen2;
	return lc;
}

ivlincomb* mult_schubert_str(const ivector* str1, const ivector* str2)
{
	assert(str_iscompat(str1, str2));

	iv_ptr dv{str2dimvec(str1)};
	if (!dv) return nullptr;
	iv_ptr w1{string2perm(str1)};
	iv_ptr w2{string2perm(str2)};

	ivlc_ptr lc{mult_schubert(w1.get(), w2.get(), int(iv_length(w1)))};
	if (!lc) return nullptr;

	w1.reset();
	w2.reset();

	ivlc_ptr res = ivlc_create();
	for (const auto& kv : ivlc_iterator(lc))
	{
		auto str = perm2string(kv.key, dv.get());
		ivlc_insert(res.get(), str, iv_hash(str), kv.value);
	}

	return res.release();
}

File: src/lrcalc/schublib.rs
use super::helper::ivector::{iv_hash, iv_new, IntVector};

use super::{
    ivlincomb::{
        _ivlc_insert, ivlc_add_element, ivlc_add_multiple, ivlc_new_default, LinearCombination,
        LC_FREE_ZERO,
    },
    perm::{
        bruhat_zero, perm2string, perm_group, perm_length, str2dimvec, str_iscompat, string2perm,
    },
};

pub fn trans(w: &[i32], vars: i32) -> LinearCombination {
    let mut res = ivlc_new_default();
    _trans(&mut w.to_vec()[..], vars, &mut res);
    res
}

fn _trans(w: &mut [i32], mut vars: i32, res: &mut LinearCombination) {
    res.clear();

    let n = perm_group(w);

    let mut r = n - 1;
    while r > 0 && w[(r - 1) as usize] < w[r as usize] {
        r -= 1;
    }
    if r <= 0 {
        let xx = iv_new(std::cmp::max(vars as u32, 1));
        let hash = iv_hash(&xx[..]);
        _ivlc_insert(res, xx, hash, 1);
        return;
    }
    vars = std::cmp::max(vars, r);

    let mut s = r + 1;
    while s < n && w[(r - 1) as usize] > w[s as usize] {
        s += 1;
    }

    let wr = w[(r - 1) as usize];
    let ws = w[(s - 1) as usize];
    w[(s - 1) as usize] = wr;
    w[(r - 1) as usize] = ws;

    let mut tmp = trans(w, vars);
    for kv in tmp.map.drain() {
        let mut vec = kv.0.ptr;
        vec[(r - 1) as usize] += 1;
        let hash = iv_hash(&vec[..]);
        _ivlc_insert(res, vec, hash, kv.1);
    }

    let mut last = 0;
    let vr = w[(r - 1) as usize];
    for i in (1..r).rev() {
        let vi = w[(i - 1) as usize];
        if last < vi && vi < vr {
            last = vi;
            w[(i - 1) as usize] = vr;
            w[(r - 1) as usize] = vi;
            _trans(w, vars, &mut tmp);
            ivlc_add_multiple(res, 1, &mut tmp, LC_FREE_ZERO);
            w[(i - 1) as usize] = vi;
        }
    }

    w[(s - 1) as usize] = ws;
    w[(r - 1) as usize] = wr;
}

fn _monk_add(i: u32, slc: &LinearCombination, rank: i32, res: &mut LinearCombination) {
    let mut add = |vec: Vec<i32>, c: i32| {
        let vec: IntVector = vec.into();
        let hash = iv_hash(&vec[..]);
        ivlc_add_element(res, c, vec, hash, LC_FREE_ZERO)
    };
    for kv in slc.map.iter() {
        let w = &(kv.0.ptr)[..];
        let c = *kv.1;
        let n = w.len() as u32;
        let wi = if i <= n {
            w[(i - 1) as usize]
        } else {
            i as i32
        };

        if i <= n + 1 {
            let mut last = 0;
            let ulen = std::cmp::max(i, n);
            for j in (1..i).rev() {
                if last < w[(j - 1) as usize] && w[(j - 1) as usize] < wi {
                    last = w[(j - 1) as usize];
                    let mut vec = vec![0; ulen as usize];
                    vec[0..n as usize].copy_from_slice(&w[..n as usize]);
                    for t in n..ulen {
                        vec[t as usize] = (t + 1) as i32;
                    }
                    vec[(j - 1) as usize] = wi;
                    vec[(i - 1) as usize] = last;
                    add(vec, -c);
                }
            }
        } else {
            let mut vec = vec![0; i as usize];
            vec[0..n as usize].copy_from_slice(&w[..n as usize]);
            for t in n..(i - 2) {
                vec[t as usize] = (t + 1) as i32;
            }
            vec[(i - 2) as usize] = i as i32;
            vec[(i - 1) as usize] = i as i32 - 1;
            add(vec, -c);
        }

        if i > n {
            let mut vec = vec![0; (i + 1) as usize];
            vec[0..n as usize].copy_from_slice(&w[..n as usize]);
            for t in n..i {
                vec[t as usize] = (t + 1) as i32;
            }
            vec[(i - 1) as usize] = i as i32 + 1;
            vec[i as usize] = i as i32;
            add(vec, c);
        } else {
            let mut last = i32::MAX;
            for j in (i + 1)..=n {
                if wi < w[(j - 1) as usize] && w[(j - 1) as usize] < last {
                    last = w[(j - 1) as usize];
                    let mut vec = w[..n as usize].to_vec();
                    vec[(i - 1) as usize] = last;
                    vec[(j - 1) as usize] = wi;
                    add(vec, c);
                }
            }
            if last > n as i32 && (n as i32) < rank {
                let mut vec = vec![0; (n + 1) as usize];
                vec[0..n as usize].copy_from_slice(&w[..n as usize]);
                vec[(i - 1) as usize] = n as i32 + 1;
                vec[n as usize] = wi;
                add(vec, c);
            }
        }
    }
}

struct Poly {
    key: IntVector,
    val: i32,
}

pub fn mult_poly_schubert(poly: &mut LinearCombination, perm: &mut IntVector, mut rank: i32) {
    let n = poly.map.len();
    if n == 0 {
        poly.clear();
        return;
    }

    if rank == 0 {
        rank = i32::MAX;
    }

    let mut p = Vec::with_capacity(n as usize);
    let mut maxvar = 0;
    for kv in poly.map.drain() {
        let mut xx = kv.0.ptr;
        let mut j = xx.length;
        while j > 0 && xx[(j - 1) as usize] == 0 {
            j -= 1;
        }
        xx.length = j;
        if maxvar < j {
            maxvar = j;
        }
        p.push(Poly { key: xx, val: kv.1 });
    }
    debug_assert!(p.len() == n as usize);

    let svlen = perm.length;
    perm.length = perm_group(&perm[..]) as u32;
    _mult_ps(&mut p[..], n as u32, maxvar, perm, rank, poly);
    perm.length = svlen;
}

fn _mult_ps(
    poly: &mut [Poly],
    n: u32,
    maxvar: u32,
    perm: &IntVector,
    rank: i32,
    res: &mut LinearCombination,
) {
    if maxvar == 0 {
        let w = perm.clone();
        let hash = iv_hash(&w[..]);
        ivlc_add_element(res, poly[0].val, w, hash, LC_FREE_ZERO);
        return;
    }

    let mut mv0 = 0;
    let mut mv1 = 0;
    let mut j = 0;
    for i in 0..n {
        let xx = &mut poly[i as usize].key;
        let mut lnxx = xx.length;
        if lnxx < maxvar {
            if mv0 < lnxx {
                mv0 = lnxx;
            }
        } else {
            xx[(maxvar - 1) as usize] -= 1;
            while lnxx > 0 && xx[(lnxx - 1) as usize] == 0 {
                lnxx -= 1;
            }
            xx.length = lnxx;
            if mv1 < lnxx {
                mv1 = lnxx;
            }
            poly.swap(i as usize, j as usize);
            j += 1;
        }
    }

    let mut res1 = ivlc_new_default();
    _mult_ps(poly, j, mv1, perm, rank, &mut res1);
    _monk_add(maxvar, &res1, rank, res);

    if j < n {
        _mult_ps(&mut poly[j as usize..], n - j, mv0, perm, rank, res);
    }
}

pub fn mult_schubert(w1: &mut IntVector, w2: &mut IntVector, mut rank: i32) -> LinearCombination {
    let w1len = perm_length(&w1[..]);
    let w2len = perm_length(&w2[..]);
    let (w1, w2, w1len, w2len) = if w1len <= w2len {
        (w1, w2, w1len, w2len)
    } else {
        (w2, w1, w2len, w1len)
    };

    let svlen1 = w1.length;
    let svlen2 = w2.length;
    w1.length = perm_group(&w1[..]) as u32;
    w2.length = perm_group(&w2[..]) as u32;

    if rank == 0 {
        rank = i32::MAX;
    } else if 2 * (w1len + w2len) > rank * (rank - 1) || bruhat_zero(&w1[..], &w2[..], rank) {
        w1.length = svlen1;
        w2.length = svlen2;
        return ivlc_new_default();
    }

    let mut lc = trans(&w1[..], 0);
    mult_poly_schubert(&mut lc, w2, rank);

    w1.length = svlen1;
    w2.length = svlen2;
    lc
}

pub fn mult_schubert_str(str1: &IntVector, str2: &IntVector) -> Option<LinearCombination> {
    debug_assert!(str_iscompat(&str1[..], &str2[..]));

    // drop dv, w1, w2
    let dv = match str2dimvec(str1) {
        None => {
            return None;
        }
        Some(v) => v,
    };
    let mut w1 = string2perm(str1);
    let mut w2 = string2perm(str2);

    let len = w1.length as i32;
    let lc = mult_schubert(&mut w1, &mut w2, len);

    drop(w1);
    drop(w2);

    let mut res = ivlc_new_default();
    for kv in lc.map.iter() {
        let str = perm2string(&(kv.0.ptr)[..], &dv[..]);
        let hash = iv_hash(&str[..]);
        _ivlc_insert(&mut res, str, hash, *kv.1);
    }

    Some(res)
}
